### ISSUE TITLE
Godot GDScript performance issues with lsp-mode

### ISSUE THREAD

**Issue by NathanLovato** [emacs-lsp/lsp-mode#2886]
When writing GDScript code with lsp-mode, errors keep appearing as you type and the autocompletion popup can take a lot of time to appear.

We did some troubleshooting thanks to @yyoncho on Discord.

What we tried:

- Checked `lsp-doctor`, everything is OK.
- Profiling the completion with Doom Emacs. With and without the popup layer. Nothing in the report explains the issue.
- Profiling with vanilla Emacs and a minimum configuration based on `lsp-start-plain.el` with `gdscript-mode`. The performance issues persist. Unfortunately but there is still nothing relevant in the profiler report.

This problem happens specifically in Emacs. There are two other GDScript LSP clients for Atom and VSCode which don't have these performance issues. Here's the VSCode client: https://github.com/godotengine/godot-vscode-plugin.

## Steps to reproduce

Godot's GDScript is an embedded, domain-specific language. The language server runs when you open a project in the engine and its editor.

So you need to download and run Godot (no install required) to use the language server. You need to install `gdscript-mode` to add support for the GDScript programming language on Emacs's side. It's available on MELPA.

Here are all the steps and a minimum test project to reproduce the issue.

I could reproduce the issue with both Emacs 27.1 and an emacs 28.0.50 with native compilation.

The steps:

- [Download Godot](https://godotengine.org/download).
- Download the minimal test project and extract the archive: [godot-lsp-test.zip](https://github.com/emacs-lsp/lsp-mode/files/6529134/godot-lsp-test.zip).
- Import and edit the project in Godot. To do so:
    - Run the Godot executable. This will open the project manager.
    - Click the Import button on the right.
    - Browse to the minimal test project folder and double-click the `project.godot` file it contains.
    - Click Import and Edit.
- From the minimal test project, open `ShakingCamera.gd` in Emacs.
- Run M-x gdscript-mode.
- Turn on lsp and import the project folder.
- After the last line inside the `_ready()` function, type `timer.wait_time = 12.0`.

The completion takes time to appear, and linter error messages get reported as you type (it can take some time for errors to appear). 

![timer-example](https://user-images.githubusercontent.com/12694995/119281059-5d31f680-bbf1-11eb-921b-a58da121f092.png)

Expected behavior: the completion popup appears instantly and updates as you type.

## More information about how Godot works

When coding from within the Godot editor, Godot has an idle time setting, after which it will recompile the script, set to 2 seconds by default.

The language requires recompiling a script to report errors and lint your code. Normally, this only happens when you pause and wait for the compiler to rerun.

Looking at the reported errors in Emacs, it looks like the language server is trying to constantly recompile the script as you type each character.

The thing that is strange is that it happens with LSP mode but not in VSCode or Atom, suggesting the issue could be linked to the client.

I'm at your disposal if you need any more information.

**Comment by yyoncho** [comment]
I tested with both my config and with vanilla and the whole compoletion process feels very lightweight and instant without slowdown or anything... 


https://user-images.githubusercontent.com/13259670/119545646-b89fe800-bd9b-11eb-807a-4d4776d75b85.mp4

Can you test without lsp-ui?

**Comment by yyoncho** [comment]
can you also check if it reproduces with only one imported project?

Can you test with 

```
  (setq create-lockfiles nil)
  (setq make-backup-files nil)
```

**Comment by NathanLovato** [comment]
Yes, I have the issue after removing all known lsp projects and importing just one. I also set the variables as requested.

The performance issues can really scale with the project and script. Sometimes it completes well for a few lines before the issue starts to appear.

Here's a recording working in a session with lsp-start-plain.el and on a much more complex project:

https://user-images.githubusercontent.com/12694995/119555513-2d066700-bd5b-11eb-8c59-0d598f4cb5f1.mp4

You can find the project's repository here: https://github.com/GDQuest/godot-2d-tactical-space-combat

If you'd like to try it, you can open `godot/TacticalSpaceCombat.gd` in Emacs. You'll need to import and edit the `godot/project.godot` file in the Godot editor too.

**Comment by yyoncho** [comment]
> Yes, I have the issue after removing all known lsp projects and importing just one. I also set the variables as requested.

Just to make sure that we are on the same page - I meant removing all projects from the server, not from lsp-mode.

Also, I do not see the warnings - are you using the latest godot server?

**Comment by NathanLovato** [comment]
> Just to make sure that we are on the same page - I meant removing all projects from the server, not from lsp-mode.

I tried clearing all known Godot projects on Godot's side, along with the cache, same issue.

> are you using the latest godot server?

Yes, running Godot 3.3.2 at the moment.

**Comment by yyoncho** [comment]
Can you set lsp-print-io to t before starting the project and then collect the logs when performing the slow completion? (Just bear in mind that lsp-print-io = t will slow down lsp-mode even more).

**Comment by yyoncho** [comment]
I will try to implement the missing methods from gd server and also check the vscode client server logs.

**Comment by NathanLovato** [comment]
Thanks much. If you have any questions about the VSCode extension, you can ask my teammate @Razoric480 as he now maintains it. He also worked on fixing issues with the GDScript language server recently.

**Comment by NathanLovato** [comment]
Here is the log after a few moments typing. It's really big, 18MB, so I zipped it.

[lsp-log.zip](https://github.com/emacs-lsp/lsp-mode/files/6541988/lsp-log.zip)

**Comment by yyoncho** [comment]
On my system I do not see any difference between vscode performance and emacs performance. lsp-start-plain.el is not a optimal configuration so it was bad idea to test performance with it(the slowdown might have been caused by bad gc-cons-treshold).

Can you try this config:

https://github.com/yyoncho/lsp-mode/commit/8967de8c18f9b60d562a5b43550fbcace6696e3f

Start it like this:

`emacs -q -l gdscript.el`

**Comment by NathanLovato** [comment]
Thanks, I just tested it and unfortunately it's still the same.

Note that I followed the docs and section on performance long before raising the issue. Actually, I've had the problem for over a year but just worked without lsp for Godot - the language server had its issues too.

**Comment by yyoncho** [comment]
I will create gitpod reproducer for this. I suspect that there is something in your env which is making client/server communication slow.
